### PR TITLE
Implement https://forum.nim-lang.org/t/7848#50018

### PR DESCRIPTION
**Pull request by c-blake**
Not sure if a 2 line output will mess up any CI things.  Let's see. ;-)

If so then we probably need an extra `Hint[BuildMode]`.  If not, we might also want `proj:..out:` on their own line as well.

Output for me looks like:
```
....
Hint:  [Link]
Hint: 80725 lines; 1.712s; 146.969MiB peakmem; proj: <Path>; out: <FullPath>
***SLOW, DEBUG BUILD***; -d:release makes code run faster. [SuccessX]
```

### CODE REVIEW

**Comment by c-blake** [comment]
Looks like `testament/testament.nim` has a line-oriented `isSuccess` that flags the first of the two lines as a failure.

I know @Araq recently added an ASCII-31 unit separator character to frame messages more precisely.  So, maybe there are already plans to make `inSuccess` the `readLine` there more robust?  Maybe some `readAll().split`?

Alternately, we could subdivide `Hint[SuccessX]`.  Or, I suppose, just do one longline with the more attention getting spelling, but that looks pretty ugly.

Recommendations?

**Comment by c-blake** [comment]
Well, I just went with a new hint category.  That seemed simplest and, really, it is a different category of message anyway, except that `SuccessX` is kind of an indiscriminate catch all.

File: tests/sets/tsets.nim [hunk 1]
```diff
@@ -173,7 +173,7 @@ type
     warnProveInit, warnProveField, warnProveIndex, warnGcUnsafe, warnGcUnsafe2,
     warnUninit, warnGcMem, warnDestructor, warnLockLevel, warnResultShadowed,
     warnUser,
-    hintSuccess, hintSuccessX,
+    hintSuccess, hintSuccessX, hintBuildMode,
```
Comment (timotheecour) [inline tests/sets/tsets.nim]: remove this diff, unrelated (refs https://github.com/nim-lang/Nim/pull/17876 which will avoid this in future)

Comment (c-blake) [reply tests/sets/tsets.nim]: done.

File: tests/ccgbugs/twrong_setconstr.nim [hunk 2]
```diff
@@ -113,7 +113,7 @@ type
     warnProveInit, warnProveField, warnProveIndex, warnGcUnsafe, warnGcUnsafe2,
     warnUninit, warnGcMem, warnDestructor, warnLockLevel, warnResultShadowed,
     warnUser,
-    hintSuccess, hintSuccessX,
+    hintSuccess, hintSuccessX, hintBuildMode,
```
Comment (timotheecour) [inline tests/ccgbugs/twrong_setconstr.nim]: ditto here in case you need to make another change to this PR, remove this diff, otherwise keep it and it'll get removed with https://github.com/nim-lang/Nim/pull/17876

File: testament/testament.nim [hunk 3]
```diff
@@ -107,7 +107,8 @@ proc isSuccess(input: string): bool =
   # that may appear in user config (eg: `--filenames`).
   # Passing `XDG_CONFIG_HOME= testament args...` can be used to ignore user config
   # stored in XDG_CONFIG_HOME, refs https://wiki.archlinux.org/index.php/XDG_Base_Directory
-  input.startsWith("Hint: ") and input.endsWith("[SuccessX]")
+  input.startsWith("Hint: ") and
```
Comment (timotheecour) [inline testament/testament.nim:110]: why is this diff needed?
`callNimCompiler` checks each line for `isSuccess`

File: doc/nimc.rst [hunk 4]
```diff
@@ -102,6 +102,7 @@ Source                           The source line that triggered a diagnostic
                                  message.
 StackTrace
 Success, SuccessX                Successful compilation of a library or a binary.
+BuildMode                        Kind of build: debug, release, danger
```
Comment (timotheecour) [inline doc/nimc.rst:105]: pre-existing problem so ignore for this PR but in future work we should have a single source of truth/documentation and keep it DRY; this is already out of sync with source code

File: testament/testament.nim (hunk 3 above)
Comment (c-blake) [reply testament/testament.nim:110]: A [BuildMode] ending line is a new kind of [SuccessX] line is the simple way to say it.

Comment (c-blake) [reply testament/testament.nim:110]: I actually mentioned this & alternatives in my 2nd comment on this ~~thread~~ PR.

**Comment by timotheecour** [comment]
how about the following instead:
* add a `hintBeginner` aimed at beginners which will show helpful hints for people new to the language, and that advanced users would silence because they're noise once you've seen them. Example messages enabled by `hintBeginner` could include this notice regarding debug build, and others.

For eg, I'd still want to see a 1-liner output summary for hintSuccessX (and possibly enhance it with which gc was used, which isn't obvious depending on user config), but I'd probably want to turn off hintBeginner, whereas with buildMode, i'd have to keep both (and have 2 lines of output instead of 1, which is annoying)
* instead of this message `-d:release makes code run faster`, IMO we should show a link to docs explaining optimization flags (because `-d:danger` is likely faster, and other options affect performance like `--passc:-flto`, `--gc:arc` etc); the msg could be: `-d:release makes code run faster, or see <link> for even faster options`

**Comment by c-blake** [comment]
Build mode is really very impactful, not only on performance but for semantics like "danger",  even for advanced users.  Advanced users can turn it of as this PR stands if they always know.  I almost always know and I'm inclined to leave it on, actually.

In general, most people want fine-grained controls like the existing hint space.. `ConvFromXtoItselfNotNeeded` is very specific as are some others.  I do not see a problem with "meta hint classes" where hints x, y, z can be referred to as a set like beginner.  I think that would best be done as some other PR, though. In that sure BuildMode might be in the beginner hint class.

I also agree that more compilation meta-data like gc-mode, threads modes, what 'cc' you are using, and maybe all defines (and even their origins) could be helpful, but that seems like another PR, too, and maybe a whole meta-category.  I'm really just trying to maybe get this one very long-standing problem maybe solved.

I think the asterisks and all caps SLOW, DEBUG build is what matters most.  As to the exact content of the follow-on message, I went with what @Araq mentioned in the Forum thread with the English grammar tightened up a bit.  You cannot cram all "see also"s into one line.  In some version I had something like "see docs for optimization" or something (on the Forum thread), but people on the Forum seemed to want to explicitly mention `-d:release`.  EDIT: and URIs can be long and also seem like "distracting noise" that makes people ignore messages.  So, ending with that may undermine this attempted solution.

**Comment by timotheecour** [comment]
> I do not see a problem with "meta hint classes" where hints x, y, z can be referred to as a set like beginner.

good idea (and yes, to be discussed in other PR)

> Build mode is really very impactful, not only on performance but for semantics like "danger", even for advanced users. 

ok, but can we make it a bit more useful, and make it show other impactful flags like which gc was used (especially given that `--gc:refc` is implicitly inferred so showing what gc was used will make things more discoverable), and possibly more in future work (eg `--threads`). So, could BuildMode become `BuildFlags`?

>  In some version I had something like "see docs for optimization" or something

actually i'm now realizing we don't even have an optimization section in the docs, unless i'm missing something, so you can leave out the link from this PR but in future work I'd suggest maybe adding it; no wonder people are confused. In future work we need a dedicated paragraph (linkable as a doc link) dedicated to explaining the different optimization flags/pragmas and their tradeoffs (including -d:danger vs -d:release, --passc:-flto, `{.inline.}` and how it affects cross-module inlining, `--opt:speed`, etc.

=> followup: https://github.com/timotheecour/Nim/issues/752

**Comment by c-blake** [comment]
Anyway, I think what it says after \*\*\*SLOW, DEBUG BUILD\*\*\* can be "fine tuned" after this PR is in.  I expect, once this is in nim-devel, that a wider audience than either this PR or the Forum thread will see it and then offer such feedback { and more...like "give me my attention back, you attention grabber" ;-) }.  So, it's probably a bad idea to hold up the PR with more bikeshedding on the exact text.

I would strongly recommend that you keep it under 80 terminal columns total (including the [BuildMode] part) because "wrapped text" with a lot of sub-syntax is another thing that makes people "tune out"/pay less attention which several people on the Forum in several threads complained was the problem with the old message.  The whole idea here is to try to grab enough attention to make beginners think just a little about performance expectations.

**Comment by c-blake** [comment]
I think we basically agree on nice future directions for the diagnostic system, but even adding to the existing compilation meta data with GC's, CC's, threads, etc. seems like another PR.  I didn't sign up for a whole diagnostic system overhaul. :-)  I think doing those as a "compile metadata" hint class would also be useful.  I suggest `CompileMetaData`. :-)

This one just segregates the build mode from the other SuccessX compilation metadata, and I think that's useful on its own.  I don't know how many times it's come up on the Forum, but it is probably in the many dozens by now.

I also agree some kind of centralized meta-page on compilation strategies for faster running code to link to would be useful.  https://compiling.nim-lang.org/ perhaps.  That could even fit on the BuildMode line. :-) Such a resource should at least mention PGO as I've seen 2+X speed-ups from that more than 5 times over the years.

File: compiler/commands.nim [hunk 5]
```diff
@@ -480,7 +480,7 @@ proc setCommandEarly*(conf: ConfigRef, command: string) =
   # must be handled here to honor subsequent `--hint:x:on|off`
   case conf.cmd
   of cmdRst2html, cmdRst2tex: # xxx see whether to add others: cmdGendepend, etc.
-    conf.foreignPackageNotes = {hintSuccessX}
+    conf.foreignPackageNotes = {hintSuccessX, hintBuildMode}
```
Comment (timotheecour) [inline compiler/commands.nim]: this diff doesn't seem good; rst2html and rst2tex; -d:release etc seems irrelevant for those commands

File: compiler/lineinfos.nim [hunk 6]
```diff
@@ -142,7 +143,8 @@ const
     warnUser: "$1",
     hintSuccess: "operation successful: $#",
     # keep in sync with `testament.isSuccess`
-    hintSuccessX: "${loc} lines; ${sec}s; $mem; $build build; proj: $project; out: $output",
+    hintSuccessX: "$loc lines; ${sec}s; $mem; proj: $project; out: $output",
+    hintBuildMode: "$build",
```
Comment (timotheecour) [inline compiler/lineinfos.nim]:
```suggestion
    hintBuildMode: "$1",
```
and then simplify the other code with:
```
rawMessage(conf, hintBuildMode, build)
```

File: compiler/commands.nim (hunk 5 above)
Comment (c-blake) [reply compiler/commands.nim]: The theory here was again that successx becomes either that or buildmode, but in truth this one I wasn't quite sure of and almost asked about it.

File: compiler/lineinfos.nim (hunk 6 above)
Comment (c-blake) [reply compiler/lineinfos.nim]: Done. CI checking.

File: compiler/commands.nim (hunk 5 above)
Comment (c-blake) [reply compiler/commands.nim]: Changed it back.

File: drnim/drnim.nim [hunk 7]
```diff
@@ -1210,19 +1210,21 @@ proc mainCommand(graph: ModuleGraph) =
...
       "loc", loc,
       "sec", sec,
       "mem", mem,
-      "build", build,
       "project", project,
       "output", ""
       ])
+    rawMessage(conf, hintBuildMode, [
+      "build", build,
+      ])
```
Comment (timotheecour) [inline drnim/drnim.nim]:
```suggestion
    rawMessage(conf, hintBuildMode, build)
```
(since you've changed to `hintBuildMode: "$1",`)

but, really, this should be refactored (maybe in another PR) to avoid this (pre-existing) code duplication

Comment (c-blake) [reply drnim/drnim.nim]: Done. Good catch.  I agree the duplication is suboptimal, but DrNim is also pretty recent/in progress.

**Comment by Araq** [comment]
I would have patched the existing SuccessX message instead but ok.

**Comment by timotheecour** [comment]
> I would have patched the existing SuccessX message instead but ok.

i agree, i think we should do this instead in followup PR instead of BuildMode, it's simpler

followups:
* patch SuccessX to add this message and remove hintBuildMode
* remove duplication with drnim
* as mentioned in https://github.com/nim-lang/Nim/pull/17874#issuecomment-827947162, add a paragraph containing optimization options that can easily be linked to and link to it from that msg

**Comment by c-blake** [comment]
@Araq - it's already a very long/noisy line of text (which has come up multiple times as the underlying reason for "not seeing" the mode) and I was trying to incorporate that user feedback.  

@timotheecour  - you & I seem to agree adding a URI (e.g. https://speed.nim-lang.org) as well as GC, maybe backend & CC & experimental flags and maybe defines/threads which, if kept on one line, could grow that line into the 100s of characters creating more visual noise that could negate the idea of a more standout DEBUG anyway.  Tooling (possibly outside of Nim-core) also seems to want line-at-a-time processing and it's easier to adapt to a new hint.  So, all together, I'd say adding hintBuildMode is simpler "in the large" if you want more text even though, yes, the patch/PR would be smaller to just change the rendering of "debug".

hintBuildMode also provides an easy, idiomatic answer to seasoned users to the likely "give that attention you are grabbing back to me" complaint - namely "just turn off hintBuildMode as you would any other hint".  Also, as mentioned, it's a different style of compile metadata (I think there are at least 3 - compiler perf, output config, and build mode). People like fine-grained controls in most compiler messaging.

So, my advice, follow-up-wise, would be to first do/try to do both "hint classes/groups", then extend to GC, CC, threads, maybe experimental flags, defines, etc., and *then* re-evaluate how hintBuildMode fits in that new taxonomy rather than knee-jerk rolling back.

Alternatively, you could maybe have the messages that get printed come from the nim.cfg/config.nims rather than being statically coded.  The default distributed config could spell things however is decided like SLOW, DEBUG mode, but then experienced users could hack theirs to say whatever they want.  So, much is already decided/defined in that config anyway, like what checks "release mode" means.

File: drnim/drnim.nim (hunk 7 above)
Comment (timotheecour) [reply drnim/drnim.nim]: => https://github.com/nim-lang/Nim/pull/18252
